cpp/include/cuspatial/experimental/iterator_factory.cuh

```diff
@@ -54,6 +75,74 @@ struct vec_2d_to_tuple {
   }
 };
 
+/**
+ * @internal
+ * @brief Generic to convert any iterator pointing to interleaved xy range into
+ * iterator of vec_2d.
+ *
+ * This generic version does not make use of vectorized load.
+ *
+ * @pre `Iter` has operator[] defined.
+ * @pre std::iterator_traits<Iter>::value_type is convertible to `T`.
+ */
+template <typename Iter, typename Enable = void>
+struct interleaved_to_vec_2d {
+  using T = typename std::iterator_traits<Iter>::value_type;
+  Iter it;
+  __device__ vec_2d<T> operator()(std::size_t i) { return vec_2d<T>{it[2 * i], it[2 * i + 1]}; }
+};
+
+/**
+ * @brief Specialization for raw pointers to interleaved xy range.
+ *
+ * @pre `Iter` is raw pointer
+ */
+template <typename Iter>
+struct interleaved_to_vec_2d<Iter, typename std::enable_if_t<std::is_pointer_v<Iter>>> {
+  using pointer_t = typename std::pointer_traits<Iter>::pointer;
+  using T         = typename std::remove_const_t<typename std::pointer_traits<Iter>::element_type>;
+  using T2        = typename element_to_element2<T>::type;
+  pointer_t ptr;
+
+  __device__ vec_2d<T> operator()(std::size_t i)
+  {
+    T2 const* f2it = reinterpret_cast<T2 const*>(ptr);
+    return vec_2d<T>{f2it[i].x, f2it[i].y};
+  }
+};
+
+/**
+ * @brief Specialization for thrust iterators conforming to `contiguous_iterator`.
+ *
+ * @pre `Iter` is a thrust iterator and conforms to `contiguous_iterator`.
+ */
+template <typename Iter>
+struct interleaved_to_vec_2d<
+  Iter,
+  typename std::enable_if_t<!std::is_pointer_v<Iter> && thrust::is_contiguous_iterator_v<Iter>>> {
+  using T  = typename std::iterator_traits<Iter>::value_type;
+  using T2 = typename element_to_element2<T>::type;
+  T* ptr;
+
+  interleaved_to_vec_2d(Iter it) { ptr = &thrust::raw_reference_cast(*it); }
+
+  __device__ vec_2d<T> operator()(std::size_t i)
+  {
+    T2 const* f2it = reinterpret_cast<T2 const*>(ptr);
+    return vec_2d<T>{f2it[i].x, f2it[i].y};
+  }
+};
+
+/**
+ * @internal
+ * @brief Functor to transform an index to strided index.
+ */
+struct strided_functor {
+  std::size_t _stride;
+  strided_functor(std::size_t stride) : _stride(stride) {}
+  auto __device__ operator()(std::size_t i) { return i * _stride; }
+};
```

Comment:
Perhaps make stride a template parameter?  Seems it's only used with a compile-time value. I don't see a need for a 64-bit stride here, either?

```suggestion
template <int Stride>
struct strided_functor {
  auto __device__ operator()(std::size_t i) { return i * Stride; }
};
```

Reply:
Inspired by @jrhemstad i think we can use a std::integer_constant here.